File: GYTableViewController/GYTableViewController/MainViewController.m
//
//  MainViewController.m
//  GYTableViewController
//
//  Created by 高扬 on 2018/3/25.
//  Copyright © 2018年 高扬. All rights reserved.
//

#import "MainViewController.h"

#import "RefreshTableViewController.h"
#import "FrameTableViewController.h"
#import "GapTableViewController.h"
#import "RelateTableViewController.h"
#import "AutoHeightTableViewController.h"
#import "CreateTableViewController.h"

//#import "GYTableViewCell.h"

@interface ControllerNode:NSObject

@property (nonatomic,copy) NSString *title;

@property (nonatomic,strong) Class controllerClass;

+ (instancetype)initWithTitle:(NSString *)title andClass:(Class)controllerClass;

@end

@implementation ControllerNode

+ (instancetype)initWithTitle:(NSString *)title andClass:(Class)controllerClass {
    ControllerNode *instance;
    @synchronized (self)    {
        if (instance == nil)
        {
            instance = [[self alloc] init];
            instance.title = title;
            instance.controllerClass = controllerClass;
        }
    }
    return instance;
}

@end

@interface ControllerDemoViewCell : GYTableViewCell

@end
@implementation ControllerDemoViewCell

- (void)showSubviews {
    self.textLabel.text = ((ControllerNode *)[self getCellData]).title;
}

- (BOOL)showSelectionStyle {
    return YES;
}

@end


@interface MainViewController ()

@property (nonatomic, strong) NSArray<ControllerNode *> *controllers;

@end

@implementation MainViewController

- (NSArray<ControllerNode *> *)controllers {
    if (!_controllers) {
        _controllers = @[
                         [ControllerNode initWithTitle:@"下拉刷新上拉加载示例" andClass:RefreshTableViewController.class],
                         [ControllerNode initWithTitle:@"Table位置、自定义刷新、点击跳转示例" andClass:FrameTableViewController.class],
                         [ControllerNode initWithTitle:@"Section或Cell间距示例" andClass:GapTableViewController.class],
                         [ControllerNode initWithTitle:@"Cell上下关系和选中高亮示例" andClass:RelateTableViewController.class],
                         [ControllerNode initWithTitle:@"Cell自动调整高度示例" andClass:AutoHeightTableViewController.class],
                         [ControllerNode initWithTitle:@"自定义创建TableView示例" andClass:CreateTableViewController.class],
                         ];
    }
    return _controllers;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = @"主页";
    
    self.edgesForExtendedLayout = UIRectEdgeNone;
    
    [self.tableView addSectionNode:[SectionNode initWithParams:^(SectionNode *sNode) {
                    [sNode addCellNodeByList:[CellNode dividingCellNodeBySourceArray:50 cellClass:ControllerDemoViewCell.class sourceArray:self.controllers]];
                }]];
    [self.tableView gy_reloadData];
}

- (BOOL)gy_useTableView {
    return YES;
}

- (BOOL)gy_useRefreshHeader {
    return NO;
}

- (void)didSelectRow:(GYTableBaseView *)tableView indexPath:(NSIndexPath *)indexPath {
    CellNode *cNode = [tableView getCellNodeByIndexPath:indexPath];
    ControllerNode *controllerNode = cNode.cellData;
    UIViewController *viewController = [[controllerNode.controllerClass alloc]init];
    viewController.title = controllerNode.title;
    viewController.hidesBottomBarWhenPushed = YES;
    [self.navigationController pushViewController:viewController animated:YES];
}

@end

File: GYTableViewController4Swift/GYTableViewController4Swift/MainViewController.swift
//
//  ViewController.swift
//  GYTableViewController4Swift
//
//  Created by gaoyang on 2019/2/3.
//  Copyright © 2019年 高扬. All rights reserved.
//

import UIKit

class ControllerNode: NSObject {
    
    var title:String?
    
    var controllerClass:UIViewController.Type?
    
    static func initWithTitle(_ title:String?, andClass controllerClass:UIViewController.Type?) -> ControllerNode {
        let controllerNode = ControllerNode()
        controllerNode.title = title;
        controllerNode.controllerClass = controllerClass;
        return controllerNode;
    }
}

class ControllerDemoViewCell: GYTableViewCell {
    
    override func showSubviews() {
        self.textLabel?.text = (self.getData() as? ControllerNode)?.title;
    }
    
    override func showSelectionStyle() -> Bool {
        return true
    }
    
}

class MainViewController: UIViewController {
    
    override func gy_useTableView() -> Bool {
        return true
    }
    
    override func gy_useRefreshHeader() -> Bool {
        return false
    }
    
    override func didSelectRow(_ tableView: GYTableBaseView!, indexPath: IndexPath!) {
        let cNode:CellNode = tableView.getCellNode(by: indexPath!)
        let controllerNode:ControllerNode? = cNode.cellData as? ControllerNode;
        let controllerClass:UIViewController.Type? = controllerNode?.controllerClass;
        let viewController:UIViewController? = controllerClass?.init()
        viewController?.title = controllerNode?.title;
        viewController?.hidesBottomBarWhenPushed = true;
        self.navigationController?.pushViewController(viewController!, animated: true)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.backgroundColor = UIColor.white
        self.title = "主页"
        
        self.tableView.add(SectionNode.initWithParams({ sNode in
            sNode?.addCellNode(byList: CellNode.dividingCellNode(bySourceArray: 50, cellClass: ControllerDemoViewCell.self, sourceArray: self.controllers))
        }))
        self.tableView.gy_reloadData()
    }
    
    private lazy var controllers:Array = [
        ControllerNode.initWithTitle("下拉刷新上拉加载示例", andClass: RefreshTableViewController.self),
        ControllerNode.initWithTitle("Table位置、自定义刷新、点击跳转示例", andClass: FrameTableViewController.self),
        ControllerNode.initWithTitle("Section或Cell间距示例", andClass: GapTableViewController.self),
        ControllerNode.initWithTitle("Cell上下关系和选中高亮示例", andClass: RelateTableViewController.self),
        ControllerNode.initWithTitle("Cell自动调整高度示例", andClass: AutoHeightTableViewController.self)
    ]

}

